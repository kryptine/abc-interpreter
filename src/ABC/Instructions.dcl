definition module ABC.Instructions

from StdFile import class <<<

:: ABCInstruction
	= IandIi !Int
	| IandIio !Int !Int
	| Ibuildh !String !Int
	| Ibuildho2 !String !Int !Int
	| Idup2_a !Int
	| Idup3_a !Int
	| Idup_a !Int
	| IeqC_b !Char !Int
	| IeqD_b !String !Int
	| IeqI_b !Int !Int
	| Ieq_desc !String !Int !Int
	| Ifillh !String !Int !Int
	| Ijmp !String
	| Ijmp_eqC_b !Char !Int !String
	| Ijmp_eqD_b !String !Int !String
	| Ijmp_eqI_b !Int !Int !String
	| Ijmp_false !String
	| Ijmp_neC_b !Char !Int !String
	| Ijmp_neI_b !Int !Int !String
	| Ijmp_true !String
	| Ijsr_eval !Int
	| Ipop_a !Int
	| Ipop_b !Int
	| Ipush2_a !Int
	| Ipush2_b !Int
	| Ipush3_a !Int
	| Ipush3_b !Int
	| IpushB !Bool
	| IpushB_a !Int
	| IpushC !Char
	| IpushC_a !Int
	| IpushD !String
	| IpushD_a !Int
	| IpushI !Int
	| IpushI_a !Int
	| Ipush_a !Int
	| Ipush_a2 !Int !Int
	| Ipush_args !Int !Int !Int
	| Ipush_b !Int
	| Ipush_r_args !Int !Int !Int
	| Iput_a !Int
	| Irepl_arg !Int !Int
	| Irepl_args !Int !Int
	| Irepl_r_args !Int !Int
	| Iswap_a1
	| Iswap_b1
	| Iupdate2pop_a !Int !Int
	| Iupdate2pop_b !Int !Int
	| Iupdate3pop_a !Int !Int
	| Iupdate3pop_b !Int !Int
	| Iupdate_a !Int !Int
	| Iupdate_b !Int !Int
	| Iupdatepop_a !Int !Int
	| Iupdatepop_b !Int !Int
	| Iupdates2pop_a !Int !Int !Int
	| Iupdates2pop_b !Int !Int !Int
	| Iupdates3pop_a !Int !Int !Int !Int
	| Iupdates3pop_b !Int !Int !Int !Int

	| IIns !String
	| Line !String
	// TODO: are these ever used?
	//| ABCInstruction
	//| ErrorABCInstruction

instance <<< ABCInstruction
instance <<< [a] | <<< a

isAStackInstruction :: ABCInstruction -> Bool
isBStackInstruction :: ABCInstruction -> Bool
