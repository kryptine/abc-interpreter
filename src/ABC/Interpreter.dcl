definition module ABC.Interpreter

from StdMaybe import :: Maybe

:: DeserializationSettings =
	{ heap_size  :: !Int //* Heap size for the interpreter, in bytes (default: 2M)
	, stack_size :: !Int //* Stack size for the interpreter, in bytes (default: 1M in total; 500k for A and 500k for BC stack)
	}

defaultDeserializationSettings :: DeserializationSettings

:: *SerializedGraph

:: InterpretedExpression
:: *InterpretationEnvironment

:: DeserializedValue a
	= DV_ParseError

	| DV_HeapFull
	| DV_StackOverflow // On Windows, not all stack overflows can be caught
	| DV_Halt
	| DV_IllegalInstruction
	| DV_SegmentationFault
	| DV_HostHeapFull

	| DV_Ok a

serialize_for_interpretation :: a !String !*World -> *(!Maybe SerializedGraph, !*World)
deserialize :: !DeserializationSettings !SerializedGraph !String !*World -> *(Maybe a, !*World)
deserializeStrict :: !DeserializationSettings !SerializedGraph !String !*World
	-> *(!DeserializedValue a, !*World)
get_start_rule_as_expression :: !DeserializationSettings !String !String !*World -> *(Maybe a, !*World)

graphToString :: !*SerializedGraph -> *(!.String, !*SerializedGraph)
graphFromString :: !String -> Maybe *SerializedGraph

graphToFile :: !*SerializedGraph !*File -> *(!*SerializedGraph, !*File)
graphFromFile :: !*File -> *(!Maybe *SerializedGraph, !*File)
