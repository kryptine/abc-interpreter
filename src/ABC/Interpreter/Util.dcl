definition module ABC.Interpreter.Util

from _SystemArray import class Array
from StdMaybe import :: Maybe
from symbols_in_program import :: Symbol

parse :: !{#Symbol} !String -> Maybe Int
new_parser :: !{#Symbol} -> Int
new_string_char_provider :: !String -> Int
free_char_provider_to_false :: !Int -> Bool
parse_program :: !Int !Int -> Int
free_to_false :: !Int -> Bool
create_array_ :: .Int -> u:(a v:b) | Array a b, [u<=v]
