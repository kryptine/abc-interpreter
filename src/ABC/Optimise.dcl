definition module ABC.Optimise

Start :: *World -> *World
