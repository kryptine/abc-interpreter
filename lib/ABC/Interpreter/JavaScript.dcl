definition module ABC.Interpreter.JavaScript

/**
 * This module provides ways to interact with JavaScript. All these functions
 * are supposed to be used from within a function that is wrapped by
 * `wrapInitFunction`, which ensures that they are run within the WebAssembly
 * runtime. It does not make sense to use these functions in native code.
 *
 * The argument to `wrapInitFunction` receives a reference to the JavaScript
 * component it is related to, which must be used to share Clean values with
 * JavaScript (where indicated in documentation below).
 *
 * The JavaScript interfacing with this module can be found in
 * abc-interpreter.js. The `ABCInterpreter` object must be instantiated with
 * `with_js_ffi: true` in the arguments to `instantiate`.
 */

import StdGeneric
from StdMaybe import :: Maybe
from StdOverloaded import class toString
from Data.Map import :: Map
from Text.GenJSON import :: JSONNode
from ABC.Interpreter import :: PrelinkedInterpretationEnvironment

/**
 * All impure interfacing with JavaScript is handled through the `*JSWorld`, as
 * native impure functionality is threaded through `*World`. `JSWorld` is not
 * unique to be able to instantiate Monad (`ABC.Interpreter.JavaScript.Monad`).
 * Because all functions using `JSWorld` use it in a unique way, sharing is
 * still prohibited.
 */
:: JSWorld

:: JSVal
:: JSFun :== JSVal
:: JSObj :== JSVal

generic gToJS a :: !a -> JSVal
derive gToJS Int, Bool, String, Real
derive gToJS JSVal, Maybe, [], (,), (,,), (,,,), (,,,,), JSONNode
derive gToJS PAIR, FIELD of {gfd_name}, RECORD
toJS x :== gToJS{|*|} x

:: JSRecord =: JSRecord (Map String JSVal)
derive gToJS JSRecord

//* Shorthand useful to build the argument for `jsRecord`.
(:>) infix 1 :: !String !a -> (String, JSVal) | gToJS{|*|} a

//* Shorthand to create a `JSRecord` (use `:>` for the elements).
jsRecord :: ![(String, JSVal)] -> JSRecord

/**
 * Store a Clean value in the JavaScript heap. The value must be associated to
 * a JS component (typically given by `wrapInitFunction`). When the component
 * is destroyed, the value may eventually be garbage collected. The value can
 * be retrieved using `jsGetCleanReference`.
 * @param Any Clean value.
 * @param A reference to a JS component.
 * @result A JavaScript reference to the Clean value.
 */
jsMakeCleanReference :: a !JSVal !*JSWorld -> *(!JSVal, !*JSWorld)

/**
 * Retrieve a Clean value from the JavaScript heap. The value must have been
 * shared using `jsMakeCleanReference`.
 */
jsGetCleanReference :: !JSVal !*JSWorld -> *(!Maybe b, !*JSWorld)

/**
 * Remove a Clean value from the JavaScript heap. The value must have been
 * shared using `jsMakeCleanReference` or `jsWrapFun`.
 */
jsFreeCleanReference :: !JSVal !*JSWorld -> *JSWorld

jsTypeOf :: !JSVal -> JSVal

jsIsUndefined :: !JSVal -> Bool
jsIsNull :: !JSVal -> Bool

jsValToInt :: !JSVal -> Maybe Int
jsValToBool :: !JSVal -> Maybe Bool
jsValToString :: !JSVal -> Maybe String
jsValToReal :: !JSVal -> Maybe Real

jsValToInt` :: !Int !JSVal -> Int
jsValToBool` :: !Bool !JSVal -> Bool
jsValToString` :: !String !JSVal -> String
jsValToReal` :: !Real !JSVal -> Real

/**
 * Retrieve a JavaScript iterable as a Clean list.
 * @param The JavaScript iterable.
 * @param The function to retrieve elements, called on every element of the
 *   iterable.
 * @result `Nothing` if the value is not iterable (has no `length` property) or
 *   if any of the elements could not be retrieved; otherwise `Just` with a
 *   list of the retrieved values.
 */
jsValToList :: !JSVal !(JSVal -> Maybe a) !*JSWorld -> *(!Maybe [a], !*JSWorld)

/**
 * Retrieve a JavaScript iterable as a Clean list.
 * @param The JavaScript iterable.
 * @param The function to retrieve elements, called on every element of the
 *   iterable.
 * @result A list of the retrieved values. If the JavaScript value is not
 *   iterable (has no `length` property), the empty list is returned.
 */
jsValToList` :: !JSVal !(JSVal -> a) !*JSWorld -> *(![a], !*JSWorld)

/**
 * Access properties of a JavaScript value.
 */
class (.#) infixl 3 attr :: !JSVal !attr -> JSVal

instance .# String // object access; may contain dots
instance .# Int // array access

/**
 * Retrieve a JavaScript value. This can be useful if the argument contains
 * computations which need to be shared among further usages of the result:
 *
 * ```
 * // object is a JavaScript object with a property 'prop' which requires computation
 * # (result,world) = object .# "prop" .? world
 * // continue to use result, so that object.prop is only evaluated once
 * ```
 *
 * However, if the argument does not contain computations, one can use a simple
 * Clean let: `# result = object .# "prop"`.
 *
 * @param The value to retrieve.
 * @result The retrieved value.
 */
(.?) infixl 1 :: !JSVal !*JSWorld -> *(!JSVal, !*JSWorld)

/**
 * Like `.?`, but forces a round-trip to JavaScript. When `.?` sees that a
 * `JSVal` in a normal environment would not trigger a computation, it does not
 * perform a round-trip but evaluates the expression itself. However, if core
 * JavaScript features have been overriden, this may fail. In these cases,
 * `jsForceFetch` may be used to force a round-trip.
 *
 * Importantly `jsForceFetch` is also needed to copy the result of
 * `jsSerializeOnClient` to the Clean world (as a string).
 */
jsForceFetch :: !JSVal !*JSWorld -> *(!JSVal, !*JSWorld)

/**
 * Set a JavaScript value to another value.
 * @param The value to set.
 * @param The new value.
 */
(.=) infixl 1 :: !JSVal !b !*JSWorld -> *JSWorld | gToJS{|*|} b

/**
 * This is an internal class representing types which can be sent as function
 * arguments to JavaScript:
 *
 * - `Int`, `Bool`, and `String` map to their JavaScript equivalents
 * - `JSVal` requires no conversion
 * - `Maybe a` uses the conversion for the underlying type for `Just`, otherwise `null`
 * - `()` relates to no arguments; tuples relates to lists of arguments
 */
class toJSArgs a :: !a -> {!JSVal}
instance toJSArgs Int, Bool, String, JSVal, [a] | gToJS{|*|} a, (Maybe a) | gToJS{|*|} a, (), JSRecord
instance toJSArgs (a,b) | gToJS{|*|} a & gToJS{|*|} b
instance toJSArgs (a,b,c) | gToJS{|*|} a & gToJS{|*|} b & gToJS{|*|} c
instance toJSArgs (a,b,c,d) | gToJS{|*|} a & gToJS{|*|} b & gToJS{|*|} c & gToJS{|*|} d
instance toJSArgs (a,b,c,d,e) | gToJS{|*|} a & gToJS{|*|} b & gToJS{|*|} c & gToJS{|*|} d & gToJS{|*|} e
instance toJSArgs (a,b,c,d,e,f) | gToJS{|*|} a & gToJS{|*|} b & gToJS{|*|} c & gToJS{|*|} d & gToJS{|*|} e & gToJS{|*|} f

/**
 * Call a JavaScript function and return the result.
 */
(.$) infixl 2 :: !JSFun !b !*JSWorld -> *(!JSVal, !*JSWorld) | toJSArgs b

/**
 * Call a JavaScript function and discard the result.
 */
(.$!) infixl 2 :: !JSFun !b !*JSWorld -> *JSWorld | toJSArgs b

/**
 * Use the JavaScript `new` keyword to create a new object.
 * @param The constructor name.
 * @param The constructor arguments.
 * @result The new object.
 */
class jsNew c :: !c !args !*JSWorld -> *(!JSVal, !*JSWorld) | toJSArgs args
instance jsNew String
instance jsNew JSVal

/**
 * Create an empty JavaScript object (`{}`).
 */
jsEmptyObject :: !*JSWorld -> *(!JSVal, !*JSWorld)

/**
 * Delete a JavaScript value with the `delete` keyword.
 */
jsDelete :: !JSVal !*JSWorld -> *JSWorld

/**
 * Lift a String to a JavaScript value using the global address space.
 */
jsGlobal :: !String -> JSVal

jsNull :== jsGlobal "null"
jsThis :== jsGlobal "this"
jsWindow :== jsGlobal "window"
jsDocument :== jsGlobal "document"

/**
 * Wrap a function for use in JavaScript. This allows it to be used as the
 * callback for events and the like. The function must be associated to an
 * JS component (typically given by `wrapInitFunction`). When the component is
 * destroyed, the function may eventually be garbage collected.
 * @param The function to wrap. When called, it receives the JavaScript
 *   arguments as an array in its first parameter.
 * @param The JS component to link the function to.
 * @result A reference to the shared function.
 */
jsWrapFun :: !({!JSVal} *JSWorld -> *JSWorld) !JSVal !*JSWorld -> *(!JSFun, !*JSWorld)

/**
 * Like {{`jsWrapFun`}}, but the Clean function can return a result.
 */
jsWrapFunWithResult :: !({!JSVal} *JSWorld -> *(JSVal, *JSWorld)) !JSVal !*JSWorld -> *(!JSFun, !*JSWorld)

/**
 * Wrap a function receiving a reference to a JavaScript component to one
 * matching the calling convention for the JavaScript interface (i.e.,
 * receiving an array of JavaScript values) so that it can be called using the
 * `initUI` step from `itasks-core.js`. Internally, this also sets up part of
 * the WebAssembly backend. The first argument to the wrapped function is a
 * reference to a JS component which can be used in `jsWrapFun` and
 * `jsMakeCleanReference`.
 */
wrapInitFunction :: !(JSVal *JSWorld -> *JSWorld) -> {!JSVal} -> *JSWorld -> *JSWorld

//* Serialize a graph for deserialization with `jsDeserializeGraph`.
jsSerializeGraph :: a !PrelinkedInterpretationEnvironment -> String

//* Deserialize a graph that was serialized with `jsSerializeGraph`.
jsDeserializeGraph :: !*String !*JSWorld -> *(!.a, !*JSWorld)

/**
 * Serialize a graph on the client side. The resulting value is a `String`, but
 * is by default not copied to the Clean world (as it is usually not needed
 * there). To get the string on the Clean side one has to use `jsForceFetch`:
 *
 * ```
 * # (ok,s) = jsSerializeOnClient graph
 * | not ok = abort "jsSerializeOnClient\n"
 * # (s,w) = jsForceFetch s w
 * # s = jsValToString` "" s
 * ```
 */
jsSerializeOnClient :: !a -> (!Bool,!JSVal)

/**
 * Deserialize a graph serialized with `jsSerializeOnClient`. This function
 * should be called on the server.
 * @param The serialized graph.
 * @param The environment (see `ABC.Interpreter`).
 * @result The deserialized value.
 * @result Always `0`.
 */
jsDeserializeFromClient :: !String !PrelinkedInterpretationEnvironment -> (.a, !Int)

/**
 * Load external CSS stylesheet by its URL.
 * @param The URL.
 */
addCSSFromUrl :: !String !*JSWorld -> *JSWorld

/**
 * Load an external JavaScript file by its URL.
 * @param The URL.
 * @param An optional callback function for when the script has loaded.
 */
addJSFromUrl :: !String !(Maybe JSFun) !*JSWorld -> *JSWorld

/**
 * A simple wrapper around JavaScript's `console.log`.
 * Use {{`jsTraceVal`}} to trace JavaScript values.
 * @param The value to log.
 * @param The value to return.
 */
jsTrace :: !a .b -> .b | toString a

/**
 * A simple wrapper around JavaScript's `console.log`.
 * Use {{`jsTrace`}} to trace Clean values.
 * @param The value to log.
 * @param The value to return.
 */
jsTraceVal :: !JSVal .a -> .a
