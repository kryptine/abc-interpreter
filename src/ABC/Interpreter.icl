implementation module ABC.Interpreter

import StdArray
import StdClass
import StdFile
import StdInt
import StdList
import StdMisc
import StdOrdList

import Data._Array
import Data.Either
import Data.Error
from Data.Func import `on`, on
import Data.Maybe
import System.CommandLine
import System.File
import System.FilePath
import System._Pointer

import graph_copy_with_names
import symbols_in_program

import ABC.Interpreter.Internal
import ABC.Interpreter.Util

:: *SerializedGraph =
	{ graph    :: !*String
	, descinfo :: !{#DescInfo}
	, modules  :: !{#String}
	, bytecode :: !String
	}

// The arguments are:
// - Pointer to C function;
// - Argument for function (in our case, pointer to the interpret node)
// - Pointer to rest of the finalizers (dealt with in the RTS)
:: Finalizer = Finalizer !Int !Int !Int
:: InterpretedExpression :== Finalizer
:: *InterpretationEnvironment = E.a:
	{ ie_finalizer :: !Finalizer
	, ie_snode_ptr :: !Int
	, ie_snodes    :: !*{a}
	}

serialize_for_interpretation :: a !FilePath !*World -> *(!SerializedGraph, !*World)
serialize_for_interpretation graph bcfile w
# (graph,descs,mods) = copy_to_string_with_names graph

# (bytecode,w) = readFile bcfile w
| isError bytecode = abort "Failed to read the bytecode file\n"
# bytecode = fromOk bytecode

#! (len,bytecodep) = strip_bytecode bytecode {#symbol_name di mods \\ di <-: descs}
#! bytecode = derefCharArray bytecodep len
| free_to_false bytecodep = abort "cannot happen\n"

# rec =
	{ graph    = graph
	, descinfo = descs
	, modules  = mods
	, bytecode = bytecode
	}
= (rec, w)
where
	symbol_name :: !DescInfo !{#String} -> String
	symbol_name {di_prefix_arity_and_mod,di_name} mod_a
	# prefix_n = di_prefix_arity_and_mod bitand 0xff
	# module_n = (di_prefix_arity_and_mod >> 8)-1
	# module_name = mod_a.[module_n]
	= make_symbol_name module_name di_name (min prefix_n PREFIX_D) +++ "\0"
	where
		PREFIX_D = 4

	strip_bytecode :: !String !{#String} -> (!Int, !Pointer)
	strip_bytecode bytecode descriptors = code {
		ccall strip_bytecode "sA:VIp"
	}

deserialize :: !SerializedGraph !FilePath !*World -> *(a, !*World)
deserialize {graph,descinfo,modules,bytecode} thisexe w
# (host_syms,w) = accFiles (read_symbols thisexe) w

# pgm = parse host_syms bytecode
| isNothing pgm = abort "Failed to parse bytecode\n"
# pgm = fromJust pgm
# int_syms = {#s \\ s <- getInterpreterSymbols pgm}
# int_syms = {#lookup_symbol_value d modules int_syms \\ d <-: descinfo}

# stack = malloc (IF_INT_64_OR_32 8 4 * STACK_SIZE)
# asp = stack
# bsp = stack + IF_INT_64_OR_32 8 4 * (STACK_SIZE-1)
# csp = stack + IF_INT_64_OR_32 4 2 * STACK_SIZE
# heap = malloc (IF_INT_64_OR_32 8 4 * (HEAP_SIZE+4))
# ie_settings = build_interpretation_environment
	pgm
	heap HEAP_SIZE stack STACK_SIZE
	asp bsp csp heap
# graph_node = string_to_interpreter int_syms graph ie_settings
#! (ie,_) = make_finalizer ie_settings
# ie = {ie_finalizer=ie, ie_snode_ptr=0, ie_snodes=unsafeCreateArray 1}
= (interpret ie (Finalizer 0 0 graph_node), w)
where
	getInterpreterSymbols :: !Pointer -> [Symbol]
	getInterpreterSymbols pgm = takeWhile (\s -> size s.symbol_name <> 0)
		[getSymbol i \\ i <- [0..get_symbol_table_size pgm-1]]
	where
		symbol_table = get_symbol_table pgm

		getSymbol :: !Int -> Symbol
		getSymbol i
		#! offset = symbol_table + i * IF_INT_64_OR_32 16 8 /* size of struct host_symbol */
		#! loc = derefInt offset
		#! name = derefString (derefInt (offset + IF_INT_64_OR_32 8 4))
		= {symbol_name=name, symbol_value=loc}

		get_symbol_table_size :: !Pointer -> Int
		get_symbol_table_size pgm = code {
			ccall get_symbol_table_size "p:I"
		}

		get_symbol_table :: !Pointer -> Pointer
		get_symbol_table pgm = code {
			ccall get_symbol_table "p:p"
		}

	string_to_interpreter :: !{#Int} !String !Pointer -> Pointer
	string_to_interpreter symbol_values graph ie = code {
		ccall string_to_interpreter "ASp:p"
	}

	lookup_symbol_value :: !DescInfo !{#String} !{#Symbol} -> Int
	lookup_symbol_value {di_prefix_arity_and_mod,di_name} mod_a symbols
		# prefix_n = di_prefix_arity_and_mod bitand 0xff
		# module_n = (di_prefix_arity_and_mod >> 8)-1
		# module_name = mod_a.[module_n]
		| prefix_n<PREFIX_D
			# symbol_name = make_symbol_name module_name di_name prefix_n
			# symbol_value = get_symbol_value symbol_name symbols
			| prefix_n<=1
				| symbol_value== -1
					= abort ("lookup_desc_info not found "+++symbol_name+++"\n")
					= symbol_value
				| symbol_value== -1
					= abort ("lookup_desc_info not found "+++symbol_name+++"\n")
					= symbol_value+2
			# symbol_name = make_symbol_name module_name di_name PREFIX_D
			# symbol_value = get_symbol_value symbol_name symbols
			| symbol_value== -1
				= abort ("lookup_desc_info not found "+++symbol_name+++"\n")
				# arity = prefix_n - PREFIX_D
				= symbol_value+(arity*8*2)+2
	where
		PREFIX_D = 4

STACK_SIZE :== (512 << 10) * 2
HEAP_SIZE :== 2 << 20

get_start_rule_as_expression :: !FilePath !*World -> *(a, *World)
get_start_rule_as_expression filename w
# ([prog:_],w) = getCommandLine w
# (syms,w) = accFiles (read_symbols prog) w
# (bc,w) = readFile filename w
| isError bc = abort "Failed to read the file\n"
# bc = fromOk bc
# pgm = parse syms bc
| isNothing pgm = abort "Failed to parse program\n"
# pgm = fromJust pgm
# stack = malloc (IF_INT_64_OR_32 8 4 * STACK_SIZE)
# asp = stack
# bsp = stack + IF_INT_64_OR_32 8 4 * (STACK_SIZE-1)
# csp = stack + IF_INT_64_OR_32 4 2 * STACK_SIZE
# heap = malloc (IF_INT_64_OR_32 8 4 * HEAP_SIZE)
# ie_settings = build_interpretation_environment
	pgm
	heap HEAP_SIZE stack STACK_SIZE
	asp bsp csp heap
# start_node = build_start_node ie_settings
#! (ie,_) = make_finalizer ie_settings
# ie = {ie_finalizer=ie, ie_snode_ptr=0, ie_snodes=unsafeCreateArray 1}
= (interpret ie (Finalizer 0 0 start_node), w)
	// Obviously, this is not a "valid" finalizer in the sense that it can be
	// called from the garbage collector. But that's okay, because we don't add
	// it to the finalizer_list anyway. This is just to ensure that the first
	// call to interpret gets the right argument.

build_interpretation_environment :: !Pointer !Pointer !Int !Pointer !Int !Pointer !Pointer !Pointer !Pointer -> Pointer
build_interpretation_environment pgm heap hsize stack ssize asp bsp csp hp = code {
	ccall build_interpretation_environment "ppIpIpppp:p"
}

build_start_node :: !Pointer -> Pointer
build_start_node ie = code {
	ccall build_start_node "p:p"
}

make_finalizer :: !Int -> (!.Finalizer,!Int)
make_finalizer ie_settings = code {
	push_finalizers
	ccall get_interpretation_environment_finalizer ":p"
	push_a_b 0
	pop_a 1
	build_r e__system_kFinalizer 0 3 0 0
	pop_b 3
	set_finalizers
	pushI 0
}

graphToFile :: !*SerializedGraph !*File -> *(!*SerializedGraph, !*File)
graphToFile {graph,descinfo,modules,bytecode} f
# (graph_cpy,graph,graph_size) = copy graph
# f = f <<< graph_size
# f = f <<< {#c \\ c <- graph_cpy}

# f = f <<< size descinfo
# f = writeArray (\di f -> f <<< di.di_prefix_arity_and_mod <<< size di.di_name <<< di.di_name) descinfo (size descinfo-1) f

# f = f <<< size modules
# f = writeArray (\m f -> f <<< size m <<< m) modules (size modules-1) f

# f = f <<< size bytecode
# f = f <<< bytecode

= ({graph=graph,descinfo=descinfo,modules=modules,bytecode=bytecode},f)
where
	copy :: !*(b a) -> *(![a], !*b a, !Int) | Array b a
	copy arr
	# (s,arr) = usize arr
	# (cpy,arr) = copy (s-1) arr []
	= (cpy,arr,s)
	where
		copy :: !Int !*(b a) ![a] -> *(![a], !*b a) | Array b a
		copy -1 arr cpy = (cpy, arr)
		copy i  arr cpy
		# (x,arr) = arr![i]
		= copy (i-1) arr [x:cpy]

	writeArray :: !(e *File -> *File) !(arr e) !Int !*File -> *File | Array arr e
	writeArray write xs -1 f = f
	writeArray write xs i f = writeArray write xs (i-1) (write xs.[i] f)

graphFromFile :: !*File -> *(!Either String *SerializedGraph, !*File)
graphFromFile f
# (graph,f) = readString f

# (_,descinfo_size,f) = freadi f
# (descinfo,f) = readArray readDescInfo (descinfo_size-1) (unsafeCreateArray descinfo_size) f

# (_,modules_size,f) = freadi f
# (modules,f) = readArray readString (modules_size-1) (unsafeCreateArray modules_size) f

# (bytecode,f) = readString f

# (end,f) = fend f
| not end = (Left "EOF not found after end of graph",f)
# (err,f) = ferror f
| err = (Left "I/O error while reading graph",f)

= (Right {graph=graph,descinfo=descinfo,modules=modules,bytecode=bytecode},f)
where
	readArray :: !(*File -> *(a, *File)) !Int !*(arr a) !*File -> *(!*arr a, !*File) | Array arr a
	readArray _ -1 xs f = (xs,f)
	readArray read i xs f
	# (x,f) = read f
	= readArray read (i-1) {xs & [i]=x} f

	readDescInfo :: !*File -> *(!DescInfo, !*File)
	readDescInfo f
	# (_,prefix_arity_and_mod,f) = freadi f
	# (name,f) = readString f
	= ({di_prefix_arity_and_mod=prefix_arity_and_mod, di_name=name}, f)

	readString :: !*File -> *(!.String, !*File)
	readString f
	# (_,size,f) = freadi f
	= freads f size

malloc :: !Int -> Pointer
malloc _ = code {
	ccall malloc "I:p"
}
