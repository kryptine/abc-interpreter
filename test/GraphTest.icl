module GraphTest

import StdEnv

from Data.Func import hyperstrict
import System.OS

import ABC.Interpreter

Start w
# (graph,w) = serialize_for_interpretation graph "GraphTest.bc" w
# ((intsquare,sub5,sub3_10,sumints,rev,foldr,ap1,ap3,map,reverse_string,reverse_array,reverse_boxed_array,reverse_recarr,recarr,toInt_rec,sumtup),w) = deserialize defaultDeserializationSettings graph (IF_WINDOWS "GraphTest.exe" "GraphTest") w
= use intsquare sub5 sub3_10 sumints rev foldr ap1 ap3 map reverse_string reverse_array reverse_boxed_array reverse_recarr recarr toInt_rec sumtup
where
	use ::
		(Int -> Int)
		(Int Int Int Int Int -> Int)
		(Int Int Int -> Int)
		([Int] -> Int)
		(A.a: [a] -> [a])
		(A.a b: (a b -> b) b [a] -> b)
		((Int -> Int) -> Int)
		((Int Int Int -> Int) -> Int)
		(A.a b: (a -> b) [a] -> [b])
		(String -> String)
		({#Int} -> {#Int})
		({Char} -> {Char})
		({#TestRecord} -> {#TestRecord})
		{#TestRecord}
		(TestRecord -> Int)
		((Int,Int,(Int,Int)) -> Int)
		-> [Int]
	use intsquare sub5 sub3_10 sumints rev foldr ap1 ap3 map reverse_string reverse_array reverse_boxed_array reverse_recarr recarr toInt_rec sumtup =
		[ intsquare 6 + intsquare 1
		, sub5 (last [1..47]) 1 2 3 (square 2)
		, sub3_10 -20 -30 3
		, sumints [1,1,2,3,4,5,6,7,8]
		, last (rev [37,36..0])
		, length (last (rev [[1..i] \\ i <- [37,36..0]]))
		, ap1 (\x -> x - 5)
		, ap1 (flip (-) 5)
		, ap3 (\x y z -> 10*x + 3*y + z)
		, foldr (\x y -> x + y) 0 [1,2,3,4,5,6,7,8,1]
		, toInt (last (rev [TestA,TestB]))
		, length [c \\ c <-: reverse_string "0123456789012345678901234567890123456"]
		, length [i \\ i <-: reverse_array {#i \\ i <- [0..36]}]
		, length [c \\ c <-: reverse_boxed_array {c\\ c <- ['A'..'e']}]
		, sum [toInt x \\ x <-: recarr]
		, sum [toInt x \\ x <-: reverse_recarr arr]
		, toInt_rec {tr_a=37*37,tr_b=TestA,tr_c=False}
		, sumtup (5,10,(15,7))
		: map (\x -> if (x == 0 || x == 10) 37 42) [0,10]
		]

:: TestT = TestA | TestB
instance toInt TestT
where
	toInt TestA = 37
	toInt TestB = 42

:: TestRecord =
	{ tr_a :: !Int
	, tr_b :: TestT
	, tr_c :: !Bool
	}
instance toInt TestRecord
where
	toInt {tr_a=a,tr_b=b,tr_c=c} = if c (toInt b / a) (a / toInt b)

graph = hyperstrict
	( square
	, sub5
	, sub5 0 10
	, sumints
	, reverse
	, foldr
	, ap1
	, ap3
	, map
	, reverse_string
	, reverse_array
	, reverse_boxed_array
	, reverse_recarr
	, arr
	, toInt_rec
	, \(a,b,(c,d)) -> a + b + c + d + 0
	)
where
	reverse_string :: String -> String
	reverse_string arr = reverse_array` arr

	reverse_array :: {#Int} -> {#Int}
	reverse_array arr = reverse_array` arr

	reverse_boxed_array :: {Char} -> {Char}
	reverse_boxed_array arr = reverse_array` arr

	reverse_recarr :: {#TestRecord} -> {#TestRecord}
	reverse_recarr arr = reverse_array` arr

	toInt_rec :: TestRecord -> Int
	toInt_rec rec = toInt rec

arr :: {#TestRecord}
arr =
	{ {tr_a=40,tr_b=TestA,tr_c=False}
	, {tr_a=4,tr_b=TestA,tr_c=True}
	, {tr_a=114,tr_b=TestB,tr_c=False}
	, {tr_a=2,tr_b=TestB,tr_c=True}
	, {tr_a=200,tr_b=TestB,tr_c=False}
	}

square :: Int -> Int
square x = x * x

sub5 :: Int Int Int Int Int -> Int
sub5 a b c d e = a - b - c - d - e

sumints :: [Int] -> Int
sumints []     = 0
sumints [x:xs] = x + sumints xs

reverse :: [a] -> [a]
reverse xs = rev [] xs
where
	rev new []     = new
	rev new [x:xs] = rev [x:new] xs

foldr :: (a b -> b) b [a] -> b
foldr op e []     = e
foldr op e [x:xs] = op x (foldr op e xs)

ap1 :: (Int -> Int) -> Int
ap1 f = f 42

ap3 :: (Int Int Int -> Int) -> Int
ap3 f = f 3 2 1

map :: (a -> b) [a] -> [b]
map f [x:xs] = [f x:map f xs]
map _ []     = []

reverse_array` arr :== {arr.[i] \\ i <- [s-1,s-2..0]} where s = size arr
