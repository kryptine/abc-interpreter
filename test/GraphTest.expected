[(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37),(DV_Ok 37)]
